AsyncDisplayKit/ASImageNode+AnimatedImage.mm

```diff
@@ -0,0 +1,226 @@
+    _animatedImage = animatedImage;
+  }
+  if (animatedImage != nil) {
+    __weak ASImageNode *weakSelf = self;
+    if ([animatedImage respondsToSelector:@selector(setCoverImageReadyCallback:)]) {
+      animatedImage.coverImageReadyCallback = ^(UIImage *coverImage) {
+        [weakSelf coverImageCompleted:coverImage];
+      };
+    }
+    
+    animatedImage.playbackReadyCallback = ^{
+      [weakSelf animatedImageFileReady];
+    };
+  }
+}
+
+- (id <ASAnimatedImageProtocol>)animatedImage
+{
+  ASDN::MutexLocker l(_animatedImageLock);
+  return _animatedImage;
+}
+
+- (void)setAnimatedImagePaused:(BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  _animatedImagePaused = animatedImagePaused;
+  ASPerformBlockOnMainThread(^{
+    if (animatedImagePaused) {
+      [self stopAnimating];
+    } else {
+      [self startAnimating];
+    }
+  });
+}
+
+- (BOOL)animatedImagePaused
+{
+  ASDN::MutexLocker l(_animatedImagePausedLock);
+  return _animatedImagePaused;
+}
+
+- (void)coverImageCompleted:(UIImage *)coverImage
+{
+  BOOL setCoverImage = YES;
+  {
+    ASDN::MutexLocker l(_displayLinkLock);
+    if (_displayLink != nil && _displayLink.paused == NO) {
+      setCoverImage = NO;
+    }
+  }
+  
+  if (setCoverImage) {
+    self.image = coverImage;
+  }
+}
+
+- (void)animatedImageFileReady
+{
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [self startAnimating];
+  });
+}
+
+- (void)startAnimating
+{
+  ASDisplayNodeAssertMainThread();
+  if (ASInterfaceStateIncludesVisible(self.interfaceState) == NO) {
+    return;
+  }
+  
+  if (self.animatedImagePaused) {
+    return;
+  }
+  
+  if (self.animatedImage.playbackReady == NO) {
+    return;
+  }
+  
+#if ASAnimatedImageDebug
+  NSLog(@"starting animation: %p", self);
+#endif
+  ASDN::MutexLocker l(_displayLinkLock);
+  if (_displayLink == nil) {
+    _playHead = 0;
+    _displayLink = [CADisplayLink displayLinkWithTarget:[ASWeakProxy weakProxyWithTarget:self] selector:@selector(displayLinkFired:)];
+    _displayLink.frameInterval = self.animatedImage.frameInterval;
+    
+    [_displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
+  } else {
+    _displayLink.paused = NO;
+  }
+}
+
+- (void)stopAnimating
+{
+#if ASAnimatedImageDebug
+  NSLog(@"stopping animation: %p", self);
+#endif
+  ASDisplayNodeAssertMainThread();
+  ASDN::MutexLocker l(_displayLinkLock);
+  _displayLink.paused = YES;
+  self.lastDisplayLinkFire = 0;
+  
+  [self.animatedImage clearAnimatedImageCache];
+}
+
+- (void)visibilityDidChange:(BOOL)isVisible
+{
+  [super visibilityDidChange:isVisible];
+  
+  ASDisplayNodeAssertMainThread();
+  if (isVisible) {
+    if (self.animatedImage.coverImageReady) {
+      self.image = self.animatedImage.coverImage;
+    }
+    [self startAnimating];
+  } else {
+    [self stopAnimating];
+  }
+}
+
+- (void)__enterHierarchy
+{
+  [super __enterHierarchy];
+  [self startAnimating];
```

Comment:
I think these two methods are unnecessary.  If we do need them, we should probably comment.  Also they should probably be the variants like -didEnterHierarchy rather than the __ internal versions.

But, the intent is certainly that visibility be cleared & set appropriately at these points so I'm thinking we don't need it.